torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -0,0 +1,297 @@
+#include "torch/csrc/autograd/python_variable_indexing.h"
+
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/autograd/python_variable.h"
+#include "torch/csrc/autograd/utils/wrap_outputs.h"
+#include "torch/csrc/utils/python_compat.h"
+#include "torch/csrc/utils/python_numbers.h"
+
+#include <vector>
+
+using namespace at;
+using namespace torch::autograd::utils;
+
+extern PyObject* THPLongTensorClass;
+#ifdef WITH_CUDA
+extern PyObject* THCPLongTensorClass;
+#endif
+
+extern bool THPModule_isTensor(PyObject *obj);
+
+namespace torch { namespace autograd {
+
+Py_ssize_t THPVariable_length(PyObject* self) {
+  HANDLE_TH_ERRORS
+  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
+  if (self_.dim() == 0) {
+    return 0;
+  }
+  return (Py_ssize_t)self_.size(0);
+  END_HANDLE_TH_ERRORS_RET(-1)
+}
+
+
+// We allow indexing by integers, slices, ellipsis, None, Variables,
+// and tuples of those types. We also handle bools as if they were a
+// Variable[ByteTensor].
+
+static int64_t count_specified_dimensions(PyObject* index) {
+  // Count the number of indexed dimensions (everything but ellipsis and None)
+  int64_t count = 0;
+  auto size = PyTuple_GET_SIZE(index);
+  for (Py_ssize_t i = 0; i < size; i++) {
+    PyObject* obj = PyTuple_GET_ITEM(index, i);
+    if (THPVariable_Check(obj)) {
+      auto& var = reinterpret_cast<THPVariable*>(obj)->cdata;
+      if (var.type().scalarType() == kByte) {
+        count += var.dim();
+      }
+    } else if (obj != Py_None && obj != Py_Ellipsis) {
+      count++;
+    }
+  }
+  return count;
+}
+
+[[noreturn]]
+static void invalid_index(PyObject* obj) {
+  throw IndexError(
+    "only integers, slices (`:`), ellipsis (`...`), None and long or byte "
+    "Variables are valid indices (got %s)", Py_TYPE(obj)->tp_name);
```

Comment:
I think you should still mention that they have to be on a correct device

Reply:
There's a different error message when things aren't on the correct devices. This one is jsut about types.

(We also don't mention other restrictions such as the indices have to be in-bounds and broadcastable)